kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/SaslInspectionIT.java

```diff
@@ -150,18 +134,103 @@ void shouldNotAuthenticateWhenSameMechanismButWrongPassword_PLAIN(
         assertClientsGetSaslAuthenticationException(cluster, topic, mechanism, clientLoginModule, username, password);
     }
 
-    // TODO assert fails if no handshake done at all
-    // TODO assert fails if client not configured for SASL
-    // TODO assert that filters don't get invoked even if a client sends a metadata after getting an error after authenticate
+    /**
+     * client handshakes with PLAIN
+     * proxy and broker have PLAIN enabled
+     * client authenticated with the correct password
+     * => client should be able to produce and consume
+     */
+    @Test
+    void shouldAuthenticateWhenSameMechanism_PLAIN_withReauth(@SaslMechanism(value = "PLAIN", principals = {
+            @SaslMechanism.Principal(user = "alice", password = "alice-secret")
+    }) @BrokerConfig(name = "connections.max.reauth.ms", value = "5000") KafkaCluster cluster,
+                                                              Topic topic)
+            throws Exception {
 
-    // TODO client tries one mech (not supported by proxy), then reattempts with the commonly supported mech
-    // TODO client tries one mech (supported by proxy, but not by broker), then reattempts with the commonly supported mech
-    // TODO all these things with older api versions
+        String mechanism = "PLAIN";
+        String clientLoginModule = "org.apache.kafka.common.security.plain.PlainLoginModule";
+        String username = "alice";
+        String password = "alice-secret";
+
+        assertClientsCanAccessCluster(cluster, topic, mechanism, clientLoginModule, username, password,
+                2, 1,
+                10_000);
+    }
+
+    /**
+     * client handshakes with SCRAM-SHA-256
+     * proxy and broker have SCRAM-SHA-256 enabled
+     * client authenticated with the correct password
+     * => client should be able to produce and consume
+     */
+    @Test
+    void shouldAuthenticateWhenSameMechanism_SCRAM_SHA_512_withReauth(@SaslMechanism(value = "SCRAM-SHA-256", principals = {
+            @SaslMechanism.Principal(user = "alice", password = "alice-secret")
+    }) @BrokerConfig(name = "connections.max.reauth.ms", value = "5000") KafkaCluster cluster,
+                                                                      Topic topic)
+            throws Exception {
+
+        String mechanism = "SCRAM-SHA-256";
+        String clientLoginModule = "org.apache.kafka.common.security.scram.ScramLoginModule";
+        String username = "alice";
+        String password = "alice-secret";
+
+        assertClientsCanAccessCluster(cluster, topic, mechanism, clientLoginModule, username, password,
+                2, 2,
+                10_000);
+    }
 
-    // TODO reauth:
-    // reauth for scram mechs
-    // reauth without account => that plugins get an empty principal
-    // reauth attempt by client which didn't use >= v1 Autn req
+    /**
+     * broker has PLAIN enabled
+     * proxy has SCRAM enabled
```

Comment:
I found this language a bit interesting since 'enabled' at the proxy is a combination of a service being present and whether the user allows insecure mechanisms. Should we have one case for an insecure mechanism, and another for a mechanism that we don't have a SaslObserver for?

Maybe it should say `proxy does not support PLAIN inspection`?

Reply:
Reword done and some other tidy ups applied.